File: src/app/shell/desktop/electron/packaged-update.ts
import { closeSync, mkdtempSync, openSync, readSync, statSync } from "node:fs";
import { join } from "node:path";
import { type ChildProcess, spawn, type SpawnOptions } from "node:child_process";

import {
  pipeResponseBodyToFile,
  type DownloadProgressHandler,
} from "@freeanima/core/config/app-update/download";

const PE_MAGIC = [0x4d, 0x5a] as const; // MZ

export type SpawnLike = (
  command: string,
  args: readonly string[],
  options: SpawnOptions,
) => ChildProcess;

/** 校验已下载的 NSIS 安装包（大小 + PE 头） */
export function verifyDownloadedInstaller(filePath: string, expectedSize?: number): void {
  const stat = statSync(filePath);
  if (expectedSize != null && stat.size !== expectedSize) {
    throw new Error(`安装包大小不符: 期望 ${expectedSize} 字节, 实际 ${stat.size} 字节`);
  }
  const fd = openSync(filePath, "r");
  try {
    const head = Buffer.alloc(2);
    readSync(fd, head, 0, 2, 0);
    if (head[0] !== PE_MAGIC[0] || head[1] !== PE_MAGIC[1]) {
      throw new Error("安装包不是有效的 Windows 可执行文件");
    }
  } finally {
    closeSync(fd);
  }
}

export async function downloadInstallerToFile(
  url: string,
  dest: string,
  opts?: {
    expectedSize?: number;
    onProgress?: DownloadProgressHandler;
    fetchImpl?: typeof fetch;
  },
): Promise<void> {
  const fetchImpl = opts?.fetchImpl ?? fetch;
  const res = await fetchImpl(url, {
    headers: { "User-Agent": "freeanima-desktop-updater", Accept: "application/octet-stream" },
    redirect: "follow",
  });
  if (!res.ok || !res.body) {
    throw new Error(`下载安装包失败 HTTP ${res.status}`);
  }
  const contentLength = res.headers.get("content-length");
  const total = opts?.expectedSize ?? (contentLength != null ? Number(contentLength) : null);
  const resolvedTotal = total != null && Number.isFinite(total) && total >= 0 ? total : null;
  await pipeResponseBodyToFile(res.body as ReadableStream<Uint8Array>, dest, {
    total: resolvedTotal,
    ...(opts?.onProgress ? { onProgress: opts.onProgress } : {}),
  });
}

/** 在 temp 下创建唯一目录并返回 setup.exe 路径 */
export function createInstallerTempPath(tempRoot: string): string {
  const dir = mkdtempSync(join(tempRoot, "freeanima-desktop-update-"));
  return join(dir, "freeanima-desktop-windows-x64-setup.exe");
}

function spawnAndWaitForPid(
  spawnImpl: SpawnLike,
  command: string,
  args: readonly string[],
  options: SpawnOptions,
  timeoutMs: number,
): Promise<number> {
  return new Promise((resolve, reject) => {
    const child = spawnImpl(command, args, options);
    let settled = false;
    const finish = (fn: () => void) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      fn();
    };

    const timer = setTimeout(() => {
      finish(() => reject(new Error("启动安装程序超时")));
    }, timeoutMs);

    child.on("error", (err) => {
      finish(() => reject(new Error(`无法启动安装程序: ${err.message}`)));
    });

    child.on("spawn", () => {
      const pid = child.pid;
      if (pid == null) {
        finish(() => reject(new Error("无法启动安装程序: 未获得进程 ID")));
        return;
      }
      child.unref();
      finish(() => resolve(pid));
    });
  });
}

const DETACHED_SPAWN_OPTS: SpawnOptions = {
  detached: true,
  stdio: "ignore",
  windowsHide: true,
};

/**
 * 在 Windows 上启动 NSIS 静默安装（/S）。
 * 优先 cmd start 脱离 Electron 进程树；失败时回退直接 spawn。
 * 不在此函数内 quit 应用——由 installer.nsh --quit-for-install 关闭。
 */
export async function launchWindowsNsisInstaller(
  installerPath: string,
  opts?: { spawnImpl?: SpawnLike; timeoutMs?: number },
): Promise<number> {
  const spawnImpl = opts?.spawnImpl ?? spawn;
  const timeoutMs = opts?.timeoutMs ?? 10_000;
  const comSpec = process.env.ComSpec ?? "cmd.exe";

  try {
    return await spawnAndWaitForPid(
      spawnImpl,
      comSpec,
      ["/d", "/c", "start", '""', "/min", installerPath, "/S"],
      DETACHED_SPAWN_OPTS,
      timeoutMs,
    );
  } catch {
    return spawnAndWaitForPid(spawnImpl, installerPath, ["/S"], DETACHED_SPAWN_OPTS, timeoutMs);
  }
}

export type ApplyPackagedUpdatePayload = {
  assetUrl: string;
  expectedSize?: number;
};

export function parsePackagedUpdatePayload(payload: unknown): ApplyPackagedUpdatePayload {
  if (typeof payload === "string" && payload.startsWith("https://")) {
    return { assetUrl: payload };
  }
  if (payload && typeof payload === "object" && "assetUrl" in payload) {
    const assetUrl = (payload as { assetUrl: unknown }).assetUrl;
    if (typeof assetUrl !== "string" || !assetUrl.startsWith("https://")) {
      throw new Error("无效的安装包 URL");
    }
    const expectedSize = (payload as { expectedSize?: unknown }).expectedSize;
    if (
      expectedSize != null &&
      (typeof expectedSize !== "number" || !Number.isFinite(expectedSize))
    ) {
      throw new Error("无效的安装包大小");
    }
    return {
      assetUrl,
      ...(typeof expectedSize === "number" ? { expectedSize } : {}),
    };
  }
  throw new Error("无效的安装包 URL");
}

File: packages/frontend/portal/app/tauri/src-tauri/src/packaged_update.rs
//! Windows NSIS 覆盖安装（移植自 Electron packaged-update）。

use serde::Serialize;
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{Duration, Instant};
use tauri::{AppHandle, Emitter, Runtime};

const PE_MAGIC: [u8; 2] = [0x4d, 0x5a]; // MZ
const PROGRESS_EVENT: &str = "shell:packaged-update-progress";
const INSTALLER_NAME: &str = "freeanima-desktop-windows-x64-setup.exe";

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PackagedUpdateProgress {
  pub received: u64,
  pub total: Option<u64>,
  pub phase: &'static str,
}

fn emit_progress<R: Runtime>(app: &AppHandle<R>, progress: PackagedUpdateProgress) {
  let _ = app.emit(PROGRESS_EVENT, progress);
}

pub fn create_installer_temp_path(temp_root: &Path) -> Result<PathBuf, String> {
  let dir = tempfile_dir(temp_root)?;
  Ok(dir.join(INSTALLER_NAME))
}

fn tempfile_dir(temp_root: &Path) -> Result<PathBuf, String> {
  let stamp = std::time::SystemTime::now()
    .duration_since(std::time::UNIX_EPOCH)
    .map(|d| d.as_millis())
    .unwrap_or(0);
  let dir = temp_root.join(format!("freeanima-desktop-update-{stamp}"));
  fs::create_dir_all(&dir).map_err(|e| format!("创建临时目录失败: {e}"))?;
  Ok(dir)
}

pub fn verify_downloaded_installer(path: &Path, expected_size: Option<u64>) -> Result<(), String> {
  let meta = fs::metadata(path).map_err(|e| format!("读取安装包失败: {e}"))?;
  if let Some(want) = expected_size {
    if meta.len() != want {
      return Err(format!(
        "安装包大小不符: 期望 {want} 字节, 实际 {} 字节",
        meta.len()
      ));
    }
  }
  let mut file = File::open(path).map_err(|e| format!("打开安装包失败: {e}"))?;
  let mut head = [0u8; 2];
  file
    .read_exact(&mut head)
    .map_err(|e| format!("读取安装包头失败: {e}"))?;
  if head != PE_MAGIC {
    return Err("安装包不是有效的 Windows 可执行文件".into());
  }
  Ok(())
}

fn download_installer_to_file<R: Runtime>(
  app: &AppHandle<R>,
  url: &str,
  dest: &Path,
  expected_size: Option<u64>,
) -> Result<(), String> {
  let client = reqwest::blocking::Client::builder()
    .user_agent("freeanima-desktop-updater")
    .redirect(reqwest::redirect::Policy::limited(10))
    .connect_timeout(Duration::from_secs(30))
    // 大安装包 + 慢网；超时后失败回传 UI，避免 Toast 永久「下载中」。
    .timeout(Duration::from_secs(600))
    .build()
    .map_err(|e| format!("创建下载客户端失败: {e}"))?;

  let mut res = client
    .get(url)
    .header(reqwest::header::ACCEPT, "application/octet-stream")
    .send()
    .map_err(|e| format!("下载安装包失败: {e}"))?;

  if !res.status().is_success() {
    return Err(format!("下载安装包失败 HTTP {}", res.status()));
  }

  let content_len = res.content_length();
  let total = expected_size.or(content_len);

  if let Some(parent) = dest.parent() {
    fs::create_dir_all(parent).map_err(|e| format!("创建目录失败: {e}"))?;
  }
  let mut file = File::create(dest).map_err(|e| format!("写入安装包失败: {e}"))?;
  let mut buf = [0u8; 64 * 1024];
  let mut received: u64 = 0;
  let mut last_emit = Instant::now() - Duration::from_secs(1);

  loop {
    let n = res
      .read(&mut buf)
      .map_err(|e| format!("下载安装包失败: {e}"))?;
    if n == 0 {
      break;
    }
    file
      .write_all(&buf[..n])
      .map_err(|e| format!("写入安装包失败: {e}"))?;
    received += n as u64;
    if last_emit.elapsed() >= Duration::from_millis(100) || total == Some(received) {
      last_emit = Instant::now();
      emit_progress(
        app,
        PackagedUpdateProgress {
          received,
          total,
          phase: "downloading",
        },
      );
    }
  }
  file.flush().map_err(|e| format!("刷新安装包失败: {e}"))?;
  emit_progress(
    app,
    PackagedUpdateProgress {
      received,
      total: total.or(Some(received)),
      phase: "downloading",
    },
  );
  Ok(())
}

fn launch_windows_nsis_installer(installer_path: &Path) -> Result<u32, String> {
  let path_str = installer_path
    .to_str()
    .ok_or_else(|| "安装包路径无效".to_string())?;

  // NSIS setup 是 GUI 子系统：直接脱离进程树，勿经 cmd.exe（会弹控制台）。
  // CREATE_NO_WINDOW 与 DETACHED_PROCESS 互斥，此处不要叠前者。
  use std::os::windows::process::CommandExt;
  const DETACHED_PROCESS: u32 = 0x00000008;
  const CREATE_NEW_PROCESS_GROUP: u32 = 0x00000200;

  let child = Command::new(path_str)
    .arg("/S")
    .stdin(std::process::Stdio::null())
    .stdout(std::process::Stdio::null())
    .stderr(std::process::Stdio::null())
    .creation_flags(DETACHED_PROCESS | CREATE_NEW_PROCESS_GROUP)
    .spawn()
    .map_err(|e| format!("无法启动安装程序: {e}"))?;
  Ok(child.id())
}

/// 下载并静默启动 NSIS；不在此退出应用（由 installer hooks / --quit-for-install 关闭）。
pub fn apply_windows_packaged_update<R: Runtime>(
  app: &AppHandle<R>,
  asset_url: &str,
  expected_size: Option<u64>,
) -> Result<(), String> {
  if !asset_url.starts_with("https://") {
    return Err("无效的安装包 URL".into());
  }

  let temp_root = std::env::temp_dir();
  let dest = create_installer_temp_path(&temp_root)?;
  download_installer_to_file(app, asset_url, &dest, expected_size)?;
  verify_downloaded_installer(&dest, expected_size)?;

  emit_progress(
    app,
    PackagedUpdateProgress {
      received: expected_size.unwrap_or(0),
      total: expected_size,
      phase: "installing",
    },
  );

  let pid = launch_windows_nsis_installer(&dest)?;
  eprintln!("[packaged-update] installer launched pid={pid} path={}", dest.display());
  Ok(())
}
